File: operon/providers/registry/embedding_plugin.py
"""Embedding resource plugin for ResourceHub.

Auto-registers embedding config classes and factory handlers with operon.
"""

from operon.core.registry import REGISTRY
from operon.providers.embeddings.config import EmbeddingConfig
from operon.providers.embeddings.factory import create_embedding

_registered = False


def register():
    """Register embedding config class and factory handler."""
    global _registered
    if _registered:
        return

    REGISTRY.register(EmbeddingConfig, create_embedding)
    _registered = True


def is_registered() -> bool:
    """Check if plugin has been registered."""
    return _registered


# Auto-register on import
register()

File: rust/operonx/src/providers/registry/embedding_plugin.rs
//! Embedding plugin registration.
//!
//! Mirrors Python [`operon/providers/registry/embedding_plugin.py`](../../../../../operon/providers/registry/embedding_plugin.py).

use std::sync::Arc;

use serde_json::Value;

use super::EmbeddingResource;
use crate::core::exceptions::OperonError;
use crate::core::registry::{registry, ConfigDict, Factory, ResourceInstance};
use crate::providers::embeddings::{create_embedder, EmbeddingConfig};

/// Register the embedding category factory. Idempotent.
pub fn register() -> Result<(), OperonError> {
    if registry().get_factory("embedding").is_some() {
        return Ok(());
    }
    let factory: Factory = Arc::new(|cfg: ConfigDict| {
        let typed: EmbeddingConfig = serde_json::from_value(Value::Object(cfg))?;
        let embedder = create_embedder(typed)?;
        Ok(Arc::new(EmbeddingResource(embedder)) as ResourceInstance)
    });
    registry().register("embedding", factory, Some("EmbeddingConfig"))
}
